### PR TITLE
IGNITE-9541 Add the comparison for two general statistics "RunAll" for master in the date interval

### PR DESCRIPTION
**Pull request by zzzadruga**

### CODE REVIEW

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/ITeamcity.java [hunk 1]
```diff
@@ -62,7 +65,19 @@
...
+    };
+
+    /**
+     * @param projectId suite ID (string without spaces)
+     * @param branch
+     * @param sinceDate
+     * @param untilDate
+     * @return list of builds in historical order, recent builds coming last
+     */
+
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/ITeamcity.java]: Remove empty line.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/ITeamcity.java [hunk 2]
```diff
@@ -62,7 +65,19 @@
...
      * @return list of builds in historical order, recent builds coming last
      */
-    List<BuildRef> getFinishedBuilds(String projectId, String branch);
+    default List<BuildRef> getFinishedBuilds(String projectId, String branch) {
+        return getFinishedBuilds(projectId, branch, null, null);
+    };
+
+    /**
+     * @param projectId suite ID (string without spaces)
+     * @param branch
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/ITeamcity.java]: Missed param description.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/ITeamcity.java [hunk 3]
```diff
@@ -80,7 +97,11 @@
     CompletableFuture<List<BuildRef>> getQueuedBuilds(@Nullable String branch);
 
     default int[] getBuildNumbersFromHistory(String projectId, String branchNameForHist) {
-        return getFinishedBuilds(projectId, branchNameForHist).stream().mapToInt(BuildRef::getId).toArray();
+        return getBuildNumbersFromHistory(projectId, branchNameForHist, null, null);
+    }
+
+    default int[] getBuildNumbersFromHistory(String projectId, String branchNameForHist, Date sinceDate, Date untilDate) {
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/ITeamcity.java]: Missec javadoc.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/ITeamcity.java [hunk 4]
```diff
@@ -52,6 +54,7 @@
 public interface ITeamcity extends AutoCloseable {
 
     String DEFAULT = "<default>";
+    long DEFAULT_BUILDS_COUNT = 1000;
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/ITeamcity.java]: Missec javadoc. Need empty line before previous field.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/IgnitePersistentTeamcity.java [hunk 5]
```diff
@@ -197,7 +203,7 @@ private IgnitePersistentTeamcity(Ignite ignite, IgniteTeamcityHelper teamcity) {
     }
 
     /**
-     * @return Build history: {@link BuildRef} lists cache, 32 parts, transaactional
+     * @return Build history: {@link BuildRef} lists cache, 32 parts, transactional
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/IgnitePersistentTeamcity.java]: Missed dot at the end of sentence.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/IgnitePersistentTeamcity.java [hunk 6]
```diff
@@ -290,6 +300,26 @@ private IgnitePersistentTeamcity(Ignite ignite, IgniteTeamcityHelper teamcity) {
 
                 return mergeByIdToHistoricalOrder(persistedValue, builds);
             });
+
+        if (sinceDate != null && untilDate != null)
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/IgnitePersistentTeamcity.java]: Is it possible to get request with `sinceDate` and without `untilDate`? And vice versa?

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/IgnitePersistentTeamcity.java [hunk 7]
```diff
@@ -591,6 +625,24 @@ private void addTestOccurrencesToStat(TestOccurrences val, String normalizedBran
         });
     }
 
+    @Override public IssuesUsagesList getIssuesUsagesList(String href) {
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/IgnitePersistentTeamcity.java]: Missec javadoc.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/IgniteTeamcityHelper.java [hunk 8]
```diff
@@ -354,18 +357,39 @@ public String basicAuthToken() {
...
-            + brachFilter
-            + ",count:1000", Builds.class).getBuildsNonNull();
+            + branchFilter
+            + ",count:" + DEFAULT_BUILDS_COUNT
+            + sinceDateFilter
+            + untilDateFilter, Builds.class).getBuildsNonNull();
+    }
+
+    public String getDateYyyyMmDdTHhMmSsZ(Date date){
+
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/IgniteTeamcityHelper.java]: Redundant blank line.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/IgniteTeamcityHelper.java [hunk 9]
```diff
@@ -378,7 +402,16 @@ public Build getBuild(String href) {
     }
 
     public ProblemOccurrences getProblems(Build build) {
-        return getJaxbUsingHref(build.problemOccurrences.href, ProblemOccurrences.class);
+        if (build.problemOccurrences != null) {
+            ProblemOccurrences problemOccurrences = getJaxbUsingHref(build.problemOccurrences.href, ProblemOccurrences.class);
+
+            problemOccurrences.problemOccurrences
+                .forEach(p -> p.buildRef = build);
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/IgniteTeamcityHelper.java]: No need new line here.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/IgniteTeamcityHelper.java [hunk 10]
```diff
@@ -409,12 +444,12 @@ public ChangesList getChangesList(String href) {
...
     /** {@inheritDoc} */
-    @Override public List<BuildRef> getFinishedBuilds(String projectId,
-        String branch) {
+    public List<BuildRef> getFinishedBuilds(String projectId,
+        String branch, Date sinceDate, Date untilDate) {
         List<BuildRef> finished = getBuildHistory(projectId,
             UrlUtil.escape(branch),
             true,
-            null);
+            null, sinceDate, untilDate);
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/IgniteTeamcityHelper.java]: Move args to the new lines.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/IgniteTeamcityHelper.java [hunk 11]
```diff
@@ -409,12 +444,12 @@ public ChangesList getChangesList(String href) {
     }
 
     /** {@inheritDoc} */
-    @Override public List<BuildRef> getFinishedBuilds(String projectId,
-        String branch) {
+    public List<BuildRef> getFinishedBuilds(String projectId,
+        String branch, Date sinceDate, Date untilDate) {
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/IgniteTeamcityHelper.java]: Place all args on the same line or different lines.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/tcmodel/hist/BuildRef.java [hunk 12]
```diff
@@ -52,6 +52,8 @@
 
     @XmlAttribute public Boolean composite;
 
+    @XmlAttribute public String webUrl;
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/tcmodel/hist/BuildRef.java]: Rebase your branch on current master.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/tcmodel/result/Build.java [hunk 13]
```diff
@@ -127,4 +137,4 @@ public Triggered getTriggered() {
     public void setTriggered(Triggered triggered) {
         this.triggered = triggered;
     }
-}
+}
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/tcmodel/result/Build.java]: File should ends with a blank line.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/tcmodel/result/issues/IssueRef.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+ * Issue short version from list of build's related issues.
+ *
+ * See example of XML, e.g. here
+ * https://ci.ignite.apache.org/app/rest/latest/builds/id:1694977/relatedIssues
+ */
+public class IssueRef {
+    @XmlAttribute public String id;
+    @XmlAttribute public String url;
+
+    @Override public String toString() {
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/tcmodel/result/issues/IssueRef.java]: Missed javadoc.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/tcmodel/result/issues/IssueUsage.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+ */
+@XmlRootElement(name = "IssueUsage")
+@XmlAccessorType(XmlAccessType.FIELD)
+public class IssueUsage {
+    @XmlElement(name = "issue")
+    private IssueRef issue;
+    @XmlElement(name = "changes")
+    private ChangesList changesList;
+
+    public IssueRef getIssue(){
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/tcmodel/result/issues/IssueUsage.java]: Missed javadoc.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/tcmodel/result/issues/IssuesUsagesList.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+
+    @XmlElement Integer count;
+
+    @XmlElement String href;
+
+    public List<IssueUsage> getIssuesUsagesNonNull() {
+        return issuesUsages == null ? Collections.emptyList() : issuesUsages;
+    }
+
+    @Override public String toString() {
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/tcmodel/result/issues/IssuesUsagesList.java]: Missed javadoc.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/web/model/current/BuildStatisticsSummary.java [hunk 17]
```diff
@@ -0,0 +1,293 @@
...
+        return getProblemsStream(buildTypeId).count();
+    }
+
+    /**
+     * Problems for all snapshot-dependencies.
+     *
+     * @param teamcity Teamcity.
+     */
+    private List<ProblemOccurrence> getProblems(@Nonnull final ITeamcity teamcity, List<BuildRef> builds){
+
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/web/model/current/BuildStatisticsSummary.java]: Redundant blank line.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/web/model/current/BuildStatisticsSummary.java [hunk 18]
```diff
@@ -0,0 +1,293 @@
...
+     * @param buildTypeId Build type id (if null - for all problems).
+     */
+    private Stream<ProblemOccurrence> getProblemsStream(String buildTypeId) {
+        if (problemOccurrenceList == null)
+            return Stream.empty();
+
+        return problemOccurrenceList.stream()
+            .filter(Objects::nonNull)
+            .filter(p -> buildTypeId == null || buildTypeId.equals(p.buildRef.buildTypeId)
+            );
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/web/model/current/BuildStatisticsSummary.java]: Redundant new line.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/web/model/current/BuildStatisticsSummary.java [hunk 19]
```diff
@@ -0,0 +1,293 @@
...
+
+        return buildProblemOccurrences;
+    }
+
+    /**
+     * Short build run result (without snapshot-dependencies result).
+     *
+     * @return printable result;
+     */
+    private Map<String, Long> getShortRes(){ return getBuildTypeOccurrences(null);
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/web/model/current/BuildStatisticsSummary.java]: Move return statement to the new line.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/web/model/current/BuildStatisticsSummary.java [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
+
+        return occurrences;
+    }
+
+    /** {@inheritDoc} */
+    @Override public void setUpdateRequired(boolean update) {
+        updateRequired = update;
+    }
+
+    @Override public boolean equals(Object o) {
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/web/model/current/BuildStatisticsSummary.java]: Missed javadoc.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/web/model/current/BuildStatisticsSummary.java [hunk 21]
```diff
@@ -0,0 +1,293 @@
...
+            Objects.equals(durationPrintable, that.durationPrintable) &&
+            Objects.equals(totalProblems, that.totalProblems);
+    }
+
+    @Override public int hashCode() {
+
+        return Objects.hash(buildId, testOccurrences, problemOccurrenceList, dependenciesProblems, relatedIssues, durationPrintable,
+            totalProblems, problemsCount, isFakeStub);
+    }
+}
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/web/model/current/BuildStatisticsSummary.java]: File should ends with a blank line.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/web/model/current/BuildStatisticsSummary.java [hunk 22]
```diff
@@ -0,0 +1,293 @@
...
+            Objects.equals(testOccurrences, that.testOccurrences) &&
+            Objects.equals(problemOccurrenceList, that.problemOccurrenceList) &&
+            Objects.equals(dependenciesProblems, that.dependenciesProblems) &&
+            Objects.equals(relatedIssues, that.relatedIssues) &&
+            Objects.equals(durationPrintable, that.durationPrintable) &&
+            Objects.equals(totalProblems, that.totalProblems);
+    }
+
+    @Override public int hashCode() {
+
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/web/model/current/BuildStatisticsSummary.java]: Redundant blank line.

File: ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/web/rest/build/GetBuildTestFailures.java [hunk 23]
```diff
@@ -149,4 +153,71 @@ public TestFailuresSummary getBuildTestFails(
...
+                if (!buildsStatistic.isFakeStub)
+                    buildsStatistics.add(buildsStatistic);
+            }
+
+            return buildsStatistics;
+        }
+    }
+
+    private BuildStatisticsSummary getBuildStatisticsSummaryNoCache(String server, int buildId) {
+
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/java/org/apache/ignite/ci/web/rest/build/GetBuildTestFailures.java]: Redundant blank line.

File: ignite-tc-helper-web/src/main/webapp/comparison.html [hunk 24]
```diff
@@ -0,0 +1,286 @@
...
+            for (var k = 0; k < tOcc.length; k++) {
+                statistics[tOcc[k]][j] = map[j].testOccurrences[tOcc[k]];
+            }
+        }
+
+        var anotherNum = (num === 1) ? 2 : 1;
+        var thisMedian, anotherMedian;
+
+        for (var i = 0; i < prOcc.length; i++) {
+
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/webapp/comparison.html]: Redundant blank line.

File: ignite-tc-helper-web/src/main/webapp/css/style-1.5.css [hunk 25]
```diff
@@ -115,4 +115,33 @@ form li:after
 	clear:both;
 	display:block;
 	margin-top:10px;
+}
+.stat{
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/webapp/css/style-1.5.css]: Missed blank lines between styles.

File: ignite-tc-helper-web/src/main/webapp/statistics.html [hunk 26]
```diff
@@ -0,0 +1,213 @@
...
+    <script src="https://code.jquery.com/ui/1.12.1/jquery-ui.js"></script>
+    <script src="js/common-1.6.js"></script>
+    <script src="js/testfails-2.0.js"></script>
+    <style>
+        table {
+            width: 70%;
+            border-collapse: collapse;
+        }
+        td, th {
+
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/webapp/statistics.html]: Redundant blank line.

File: ignite-tc-helper-web/src/main/webapp/statistics.html [hunk 27]
```diff
@@ -0,0 +1,213 @@
...
+        $.ajax({ url: "rest/branches/version",  success: showVersionInfo, error: showErrInLoadStatus });
+
+        if(g_updTimer==null) {
+            g_updTimer=setTimeout(tstTimeout, 3200);
+        }
+        setInterval(tstTimeout, 10000);
+    });
+
+    function tstTimeout() {
+
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/webapp/statistics.html]: Redundant blank line.

File: ignite-tc-helper-web/src/main/webapp/statistics.html [hunk 28]
```diff
@@ -0,0 +1,213 @@
...
+    $(document).ready(function() {
+        $.getScript("js/testfails-2.0.js", function(data, textStatus, jqxhr){ });
+
+        $( document ).tooltip();
+        loadData();
+        //todo fix setInterval( function() { checkForUpdate(); }, 30000);
+
+        $.ajax({ url: "rest/branches/version",  success: showVersionInfo, error: showErrInLoadStatus });
+
+        if(g_updTimer==null) {
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/webapp/statistics.html]: Need space after `if`. Below too.

File: ignite-tc-helper-web/src/main/webapp/statistics.html [hunk 29]
```diff
@@ -0,0 +1,213 @@
...
+    });
+
+    function tstTimeout() {
+
+        if(g_updTimer!=null) {
+            clearTimeout(g_updTimer);
+            g_updTimer=null;
+        }
+
+
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/webapp/statistics.html]: Don't use 2 blank lines in a row.

File: ignite-tc-helper-web/src/main/webapp/statistics.html [hunk 30]
```diff
@@ -0,0 +1,213 @@
...
+        return res;
+
+    }
+</script>
+<br>
+<div id="loadStatus"></div>
+<div id="statistics"></div>
+<div id="version"></div>
+</body>
+</html>
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/webapp/statistics.html]: File should ends with a blank line.

File: ignite-tc-helper-web/src/main/webapp/statistics.html [hunk 31]
```diff
@@ -0,0 +1,213 @@
...
+
+            res += "</td><td>" + buildStatistics.durationPrintable + "</td></tr>";
+        }
+
+        res += "</table><br>min: " + problemCountMin + "; max: " + problemCountMax +
+            "; median: " + (Math.round(median(problemCounts) * 1000) / 1000) +
+            "; average: " + (Math.round(average / result.length * 1000) / 1000) + "<br><br>";
+
+        return res;
+
```
Comment (SomeFire) [inline ignite-tc-helper-web/src/main/webapp/statistics.html]: Redundant blank line.

**Comment by dspavlov** [comment]
I've applied PR locally and tried it on existing DB. The result is a failure by not found build
```
 'Exception UncheckedIOException: java.io.FileNotFoundException: Service https://ci.ignite.apache.org/app/rest/latest/builds/id:1362502 returned not found error.Responding with error, status code: 404 (Not Found).
Details: jetbrains.buildServer.server.rest.errors.NotFoundException: No build found by id '1362502'.
Could not find the entity requested. Check the reference is correct and the user has permissions to access the entity.
```

It is quite normal TC Bot has reference to build, but TC don't have it (history in the bot is longer that in TC), so it is up to bot correctly handle absent REST requests.

**Pull request closed without merging**